### PR TITLE
[SHELL32] SHMultiFileProperties workaround and minimal SHObjectProperties

### PR DESCRIPTION
**Pull request by whindsaks**

### CODE REVIEW

File: dll/win32/shell32/stubs.cpp [hunk 1]
```diff
@@ -136,6 +145,19 @@ WINAPI
 SHMultiFileProperties(IDataObject *pDataObject, DWORD dwFlags)
 {
     FIXME("SHMultiFileProperties() stub\n");
+
+    // Temporary workaround to display a property sheet when possible
```
Comment (JoachimHenze) [inline dll/win32/shell32/stubs.cpp:149]:
```suggestion
    // Temporary workaround to display a property sheet if possible
```
What is the condition that it is possible? And what is the condition that it is impossible?
Can you attach a screenshot or small video of the after-state? Thx.

**Review by JoachimHenze — APPROVED** [review]

File: dll/win32/shell32/stubs.cpp (hunk 1 above)
Comment (whindsaks) [reply dll/win32/shell32/stubs.cpp:149]: It is only possible if the dataobject contains a single item. It does nothing to help the case where there are multiple files.

File: dll/win32/shell32/wine/shellord.c [hunk 2]
```diff
@@ -2314,6 +2314,21 @@ INT WINAPI SHHandleUpdateImage(PCIDLIST_ABSOLUTE pidlExtra)
 
 BOOL WINAPI SHObjectProperties(HWND hwnd, DWORD dwType, LPCWSTR szObject, LPCWSTR szPage)
 {
+#ifdef __REACTOS__
```
Comment (binarymaster) [inline dll/win32/shell32/wine/shellord.c:2317]: Shell32 is forked: `dll/win32/shell32             # Forked at Wine-20071011`

**Review by learn-more — APPROVED** [review]

File: dll/win32/shell32/wine/shellord.c [hunk 3]
```diff
@@ -2314,6 +2314,20 @@ INT WINAPI SHHandleUpdateImage(PCIDLIST_ABSOLUTE pidlExtra)
 
 BOOL WINAPI SHObjectProperties(HWND hwnd, DWORD dwType, LPCWSTR szObject, LPCWSTR szPage)
 {
+    LPITEMIDLIST pidl = NULL;
+    switch (dwType)
+    {
+        case SHOP_FILEPATH: pidl = ILCreateFromPathW(szObject); break;
```
Comment (HBelusca) [inline dll/win32/shell32/wine/shellord.c:2320]:
```suggestion
        case SHOP_FILEPATH:
            pidl = ILCreateFromPathW(szObject);
            break;
```

**Review by HBelusca — APPROVED** [review]
LGTM.
Out of curiosity, do you know of any 3rd-party program/shell extension/... that makes use of these functions?

**Comment by whindsaks** [comment]
> LGTM. Out of curiosity, do you know of any 3rd-party program/shell extension/... that makes use of these functions?

I have personally called them before. I can't name a specific application off the top of my head. It does not help that the ROS shell tends to give no feedback for things that fail so you don't really know when things are failing and if ROS is to blame. The symptom would just be a button/menu that does nothing and you would have to attach a debugger to find out...

File: dll/win32/shell32/wine/shellord.c (hunk 3 above)
Comment (whindsaks) [reply dll/win32/shell32/wine/shellord.c:2320]: Does it really look better to you?

The additional types would just be `case SHOP_PRINTER: pidl = ILCreatePrinterPidl(name); break;` etc. But printers are a long way down on the shell to do list.

Comment (HBelusca) [reply dll/win32/shell32/wine/shellord.c:2320]: just our usual general code style.

File: dll/win32/shell32/stubs.cpp [hunk 4]
```diff
@@ -136,6 +145,19 @@ WINAPI
...
+    // Temporary workaround to display a property sheet if possible
+    if (DataObject_GetHIDACount(pDataObject) == 1)
+        return SHELL32_ShowPropertiesDialog(pDataObject);
+
+    if (pDataObject)
+    {
+        HWND hWnd;
+        if (FAILED(IUnknown_GetWindow(pDataObject, &hWnd))) // Will probably not work but we have no other option
+            hWnd = NULL;
+        SHELL_ErrorBox(hWnd, HRESULT_FROM_WIN32(ERROR_NOT_SUPPORTED));
```
Comment (HBelusca) [inline dll/win32/shell32/stubs.cpp:158]: Does Windows also pop up a message box in that failure case?

**Comment by HBelusca** [comment]
Here's a 3rd-party opensource shell extension that uses SHMultiFileProperties:
https://github.com/zenden2k/image-uploader/blob/644e91a22d209d3c287adc5e1ceb1a3807f793b7/Source/3rdpart/ShellPidl.cpp#L286

File: dll/win32/shell32/stubs.cpp (hunk 4 above)
Comment (whindsaks) [reply dll/win32/shell32/stubs.cpp:158]: Windows does not have that failure case.

If I pass something that is not really a valid file (the desktop or a file that does not exist, it still displays the propsheet dialog:

screenshot

If I pass a NULL dataobject, it does nothing. If I pass a dataobject with zero items, Windows displays a MessagBox that it cannot display the properties for this item.

**Pull request merged**